Create Web API 8 overview page (#262)
MicrosoftDocs/OData-docs: Odata-docs/webapi-8/overview/overview.md

---
title:  "OData Web API 8 Overview"

> description: A high-level overview of OData Web API, its benefits and common use-cases.

Reviewer: ```suggestion
description: A high-level overview of OData Web API, its benefits and common use cases.
```

date:   2022-09-05
ms.date: 9/5/2022
author: habbes
ms.author: clhabins
---

# Overview
**Applies To**:[!INCLUDE[appliesto-webapi](../../includes/appliesto-webapi-v8.md)]

> OData Web API is a .NET library for building REST API services that conform to the OData protocol. The OData protocol defines best practices for consistent and strongly-typed REST APIs by specifying the format of requests and responses, type definition, and service description as well as features like filtering, querying related data, pagination, etc. OData Web API library is a set of services that extend [ASP.NET Core](/aspnet/core/introduction-to-aspnet-core) to provide OData-based routing, request and response serialization, query-handling, and more. To learn more about OData, visit [the OData docs](/odata/overview).

Author: I went with ODat Web API for general name and OData Web API 7/8 when specifying the version. Do you have a different suggestion?

> OData Web API is a .NET library for building REST API services that conform to the OData protocol. The OData protocol defines best-practices for consistent and strongly-typed REST APIs by specifying the format of requests and responses, type definition and service description as well as features like filtering, querying related data, pagination, etc. OData Web API library is a set of services that extend [ASP.NET Core](/aspnet/core/introduction-to-aspnet-core) to provide OData-based routing, request and response serialization, query handling and more. To learn more about OData itself, visit [the OData docs](/odata/overview).

Reviewer: ```suggestion
OData Web API is a .NET library for building REST API services that conform to the OData protocol. The OData protocol defines best-practices for consistent and strongly-typed REST APIs by specifying the format of requests and responses, type definition and service description as well as features like filtering, querying related data, pagination, etc. OData Web API library is a set of services that extend [ASP.NET Core](/aspnet/core/introduction-to-aspnet-core) to provide OData-based routing, request and response serialization, query handling and more. To learn more about OData, visit [the OData docs](/odata/overview).
```

> OData Web API is a .NET library for building REST API services that conform to the OData protocol. The OData protocol defines best-practices for consistent and strongly-typed REST APIs by specifying the format of requests and responses, type definition and service description as well as features like filtering, querying related data, pagination, etc. OData Web API library is a set of services that extend [ASP.NET Core](/aspnet/core/introduction-to-aspnet-core) to provide OData-based routing, request and response serialization, query handling and more. To learn more about OData itself, visit [the OData docs](/odata/overview).

Reviewer 2: best practices

> OData Web API is a .NET library for building REST API services that conform to the OData protocol. The OData protocol defines best-practices for consistent and strongly-typed REST APIs by specifying the format of requests and responses, type definition and service description as well as features like filtering, querying related data, pagination, etc. OData Web API library is a set of services that extend [ASP.NET Core](/aspnet/core/introduction-to-aspnet-core) to provide OData-based routing, request and response serialization, query handling and more. To learn more about OData itself, visit [the OData docs](/odata/overview).

Reviewer 2: type definition, and service description

> OData Web API is a .NET library for building REST API services that conform to the OData protocol. The OData protocol defines best-practices for consistent and strongly-typed REST APIs by specifying the format of requests and responses, type definition and service description as well as features like filtering, querying related data, pagination, etc. OData Web API library is a set of services that extend [ASP.NET Core](/aspnet/core/introduction-to-aspnet-core) to provide OData-based routing, request and response serialization, query handling and more. To learn more about OData itself, visit [the OData docs](/odata/overview).

Reviewer 2: query-handling, and more.

Here are some of the features provided by the library:

> - Exposes a metadata endpoint that returns a document describing the API service' model, the data types it defines and the endpoints and capabilities it exposes. This document makes it easier for clients to consume the service. It can also be used by tooling to automatically generate client-side code that can interact with the service (e.g. [OData CLI](/odata/odatacli/getting-started), [OData Connected Service](/odata/connectedservice/getting-started)), allowing the user to focus on business logic.

Reviewer: ```suggestion
- Exposes a metadata endpoint that returns a document describing the API service' model, the data types it defines and the endpoints and capabilities it exposes. This document makes it easier for clients to consume the service. It can also be used by code generation tools to automatically generate service proxy on the clients to facilitate interaction with the service (e.g., [OData CLI](/odata/odatacli/getting-started), [OData Connected Service](/odata/connectedservice/getting-started)), allowing the user to focus on business logic.
```

> - Convention-based routing: Requests are routed to controllers and actions based on naming and URL conventions. You can easily override this using ASP.NET core's `[Route]` attribute.

Reviewer 2: ASP.NET Core's

Author: Rephrased using @gathogojr's suggestion.

  - `$filter` allows you to filter results based on specific conditions: `/Products?$filter=Price lt 1000 and Category eq 'Electronics'` will only return electronic products whose price is less than 1,000.
  - `$select` allows you to limit the fields that are included in the response: `/Products?$select=Name,Category,DateCreated` will only include the `Name`, `Category`, `DateCreated` fields in the response
  - `$expand` allows you to join and include related data in the response: `/OrderItems?$expand=Product` will return order items and include the related product in the response
  - `$skip` and `$top` allow you to limit the number of records in the response and can be used for client-driven pagination: `/Products$top=10&$skip=20&$orderby=Name` returns the third page of 10 products sorted by name.
  - `$apply` allows you to apply aggregations: `/Orders?$apply=groupby(Product/Name, aggregate($count as Count))` returns the number of orders per product name.

## Why choose OData Web API:

Here are some of the benefits and common use cases for using OData Web API:

> - Creating a standards-compliant Web API with support for filtering, selecting, expanding

Reviewer: ```suggestion
- Creating a standards-compliant Web API with support for filtering, selecting, expanding among other capabilities.
```

> 

Reviewer: You didn't leave an empty line below _Here are some of the features provided by the library:_ on the section above. Depending on whether it looks better with or without the empty line, determine whether to add an empty line on the other section or remove the empty line here - for consistency.

Author: They seem to be rendered the same way on the web page. Let me remove the empty line since I have it in fewer places.

- Interoperability with OpenAPI and Swagger
- Type-safety
- Ability to generate client code from OData model description
- Expose a REST-API layer to your data source for analytics functions.

> - Add advanced querying capabilities and via OData query options to your REST API even if it's not based on OData.

Reviewer: ```suggestion
- Add advanced querying capabilities via OData query options to your REST API even if it's not based on OData.
```

> - Expose a REST-API layer to your data source for analytics functions.

Reviewer: ```suggestion
- Expose a REST API layer to your data source for analytics functions.
```

> - Add advanced querying capabilities via OData query options to your REST API even if it's not based on OData.

Reviewer: ```suggestion
- Add advanced querying capabilities via OData query options to your REST API even when it's not based on OData.
```

## NuGet packages

The library is shipped on NuGet under the package name: `Microsoft.AspNetCore.OData`.

We also ship a library separately for building OData models from .NET model classes. This is available under the package name: `Microsoft.OData.ModelBuilder`.

## Choosing the right version

There are two major versions of OData Web API under active maintenance and support:
- OData Web API 8: `Microsoft.AspNetCore.OData` 8.x
- OData Web API 7: `Microsoft.AspNetCore.OData` 7.x and `Microsoft.AspNet.OData` 7.x

> For newer development on .NET, OData Web API 8 is recommended. However, OData Web API 7 is actively maintained and some newever features in OData Web API 8 are back-ported to OData Web API 7 where possible. There also features in Web API 7 which have not been ported to Web API 8, like the [automatic support for deeply nested paths](/odata/webapi/automatic-nested-paths-with-enable-nested-paths).

Reviewer 2: newer features

> For newer development on .NET, OData Web API 8 is recommended. However, OData Web API 7 is actively maintained and some newever features in OData Web API 8 are back-ported to OData Web API 7 where possible. There also features in Web API 7 which have not been ported to Web API 8, like the [automatic support for deeply nested paths](/odata/webapi/automatic-nested-paths-with-enable-nested-paths).

Reviewer 2: There're/There are

Author: What is this in reference to?

> For newer development on .NET, OData Web API 8 is recommended. However, OData Web API 7 is actively maintained and some newever features in OData Web API 8 are back-ported to OData Web API 7 where possible. There also features in Web API 7 which have not been ported to Web API 8, like the [automatic support for deeply nested paths](/odata/webapi/automatic-nested-paths-with-enable-nested-paths).

Reviewer: ```suggestion
For newer development on .NET, OData Web API 8 is recommended. However, OData Web API 7 is actively maintained and some newer features in OData Web API 8 are back-ported to OData Web API 7 where possible. There also features in Web API 7 which have not been ported to Web API 8, e.g., [automatic support for deeply nested paths](/odata/webapi/automatic-nested-paths-with-enable-nested-paths).
```

> For newer development of OData services on .NET, OData Web API 8 is recommended. However, OData Web API 7 is actively maintained and some newer features in OData Web API 8 are back-ported to OData Web API 7 where possible. There are also features in Web API 7 that have not been ported to Web API 8, e.g., [automatic support for deeply nested paths](/odata/webapi/automatic-nested-paths-with-enable-nested-paths).

Reviewer: ```suggestion
For newer development of OData services on .NET, OData Web API 8 is recommended. However, OData Web API 7 is actively maintained and some newer features in OData Web API 8 are back-ported to OData Web API 7 where possible. There are also features in OData Web API 7 that have not been ported to OData Web API 8, e.g., [automatic support for deeply nested paths](/odata/webapi/automatic-nested-paths-with-enable-nested-paths).
```

> For newer development of OData services on .NET, OData Web API 8 is recommended. However, OData Web API 7 is actively maintained and some newer features in OData Web API 8 are back-ported to OData Web API 7 where possible. There are also features in Web API 7 that have not been ported to Web API 8, e.g., [automatic support for deeply nested paths](/odata/webapi/automatic-nested-paths-with-enable-nested-paths).

Reviewer 3: Do we have a page with a list of features that are not back-ported?If we do we can include that list instead of a single example.  if not we may want to remove the sentence about features that are not back ported for now until we can link to a list of these features

Author: We don't have such a page in the official docs. I can create a list here, but it won't be exhaustive (since I don't know the full extent of what's not been ported to 8 or backported to 7. Maybe others can contribute.

Author: I've gotten one example from a @KenitoInc  of a feature that's not back-ported from 8 to 7 (i.e. it's only available to 8), but I don't have an exhaustive list. Not sure if you'd prefer that I mention that example or still remove the sentence for now?

Author: Actually, after re-reading the statement and your comment, I have opted to remove it for now.

> For newer development of OData services on .NET, OData Web API 8 is recommended. However, OData Web API 7 is actively maintained and some newer features in OData Web API 8 are back-ported to OData Web API 7 where possible. There are also features in Web API 7 that have not been ported to Web API 8, e.g., [automatic support for deeply nested paths](/odata/webapi/automatic-nested-paths-with-enable-nested-paths).

Reviewer 3: ```suggestion
For new development of OData services on .NET, OData Web API 8 is recommended. However, OData Web API 7 is actively maintained and some newer features in OData Web API 8 are back-ported to OData Web API 7 where possible. There are also features in Web API 7 that have not been ported to Web API 8, e.g., [automatic support for deeply nested paths](/odata/webapi/automatic-nested-paths-with-enable-nested-paths).
```

> OData Web API takes better advantage of ASP.NET Core features. However, it's only available for .NET Core 3.1 and newer versions of .NET. It does not support .NET Framework. OData Web API 7 supports ASP.NET classic and .NET Framework through the `Microsoft.AspNet.OData` package.

Reviewer: ```suggestion
OData Web API 8 leverages ASP.NET Core features better. However, it's only available for .NET Core 3.1 and newer versions of .NET. It does not support .NET Framework. OData Web API 7 supports ASP.NET classic and .NET Framework through the `Microsoft.AspNet.OData` package.
```

> OData Web API 8 leverages ASP.NET Core features better. However, it's only available for .NET Core 3.1 and newer versions of .NET. It does not support .NET Framework. OData Web API 7 supports ASP.NET classic and .NET Framework through the `Microsoft.AspNet.OData` package.

Reviewer 3: ```suggestion
OData Web API 8 leverages ASP.NET Core features and is only available for .NET Core 3.1 and newer versions of .NET. It does not support .NET Framework. OData Web API 7 supports ASP.NET classic and .NET Framework through the `Microsoft.AspNet.OData` package.
```
